File: basics/realloc/native/ts/state/enhanced-address-info.ts
import { Buffer } from 'node:buffer';
import * as borsh from 'borsh';

export class EnhancedAddressInfo {
  name: string;
  house_number: number;
  street: string;
  city: string;
  state: string;
  zip: number;

  constructor(props: {
    name: string;
    house_number: number;
    street: string;
    city: string;
    state: string;
    zip: number;
  }) {
    this.name = props.name;
    this.house_number = props.house_number;
    this.street = props.street;
    this.city = props.city;
    this.state = props.state;
    this.zip = props.zip;
  }

  toBase58() {
    return borsh.serialize(EnhancedAddressInfoSchema, this).toString();
  }

  toBuffer() {
    return Buffer.from(borsh.serialize(EnhancedAddressInfoSchema, this));
  }

  static fromBuffer(buffer: Buffer) {
    return borsh.deserialize(EnhancedAddressInfoSchema, EnhancedAddressInfo, buffer);
  }
}

export const EnhancedAddressInfoSchema = new Map([
  [
    EnhancedAddressInfo,
    {
      kind: 'struct',
      fields: [
        ['name', 'string'],
        ['house_number', 'u8'],
        ['street', 'string'],
        ['city', 'string'],
        ['state', 'string'],
        ['zip', 'u32'],
      ],
    },
  ],
]);

File: basics/realloc/pinocchio/program/src/state/enhanced_address_info.rs
pub struct EnhancedAddressInfoExtender {
    pub state: [u8; 8],
    pub zip: u32,
}

pub struct EnhancedAddressInfo {
    pub name: [u8; 8],
    pub house_number: u8,
    pub street: [u8; 8],
    pub city: [u8; 8],
    pub state: [u8; 8],
    pub zip: u32,
}

impl EnhancedAddressInfo {
    pub const LEN: usize = 37;
}
